ユーザーフォローの提出物

yocaji opened the pull request:
ユーザーフォローの提出物になります。レビューよろしくお願いいたします🙇‍♂️

- 参考にした情報
  - [第14章 ユーザーをフォローする - Railsチュートリアル](https://railstutorial.jp/chapters/following_users?version=5.0)
  - [フォロー・フォロワー機能｜Web開発で学ぶプログラミング](https://zenn.dev/odentravel/books/e69a157daeecb3/viewer/086fe9)
  - [Rails ユーザーフォロー機能を追加する -](https://yuna627.hatenablog.jp/entry/2020/12/15/054633)

Commit:
フォローModel作成

Commit:
モデルを編集した

Commit:
ルーティングを追加した

Commit:
コントローラとビューを新規作成/編集した

Commit:
不要なファイルとコメントを削除

Commit:
誤って一括変換されていた個所を修正した

maedana commented on db/schema.rb:
```diff
@@ -49,6 +49,16 @@
 
+  create_table "following_relations", force: :cascade do |t|
+    t.integer "follower_id"
+    t.integer "followed_id"
```
外部キーで両方とも必ず入れる値だと思いますのでnot null制約いれましょうー

yocaji replied:
not null制約を追加しました

maedana commented on config/routes.rb:
```diff
@@ -3,5 +3,10 @@
   root to: 'books#index'
   resources :books
-  resources :users, only: %i(index show)
+  resources :users, only: %i[index show edit update] do
```
controllerみた感じですと、edit/updateは不要なのではないでしょうか?

yocaji replied:
ご指摘の通りです。修正しました。

maedana commented on app/models/user.rb:
```diff
@@ -5,4 +5,33 @@ class User < ApplicationRecord
+  end
+
+  def unfollow(user)
+    active_following_relations.find_by(followed_id: user.id).destroy
```
`find_by`の結果見つからないときに`destroy`が呼べずエラーになりそうかなと思いました。見つからない時でもエラーにならないようにしたいところです。誰かをフォロー中の画面を複数タブ開くことで発生させることは出来そうですね

yocaji replied:
教えていただいた条件でエラーを起こして、何が起こるか理解できました🙇‍♂️
`find_by`の結果がnilの時でもエラーにならないように、という形にしました。

maedana commented on app/models/user.rb:
```diff
@@ -5,4 +5,33 @@ class User < ApplicationRecord
+           source: :follower
+
+  def follow(user)
+    active_following_relations.create(followed_id: user.id)
```
DBにユニーク制約がありますが、Modelのバリデーションがない様子なので既にフォロー中のときにこの処理走るとエラーが発生したりしないですかね?

yocaji replied:
こちらも教えていただいた条件でエラーを起こして、何が起こるか理解できました🙇‍♂️

maedana commented on app/models/following_relation.rb:
```diff
@@ -0,0 +1,6 @@
+
+class FollowingRelation < ApplicationRecord
+  belongs_to :follower, class_name: 'User'
+  belongs_to :followed, class_name: 'User'
```
DB側にユニーク制約ありますが、Modelでも`t.index %i[follower_id followed_id], unique: true`に対するvalidationを記述したほうがよさそうかなと思いました。

yocajii replied:
validationを追加してみました。

maedana commented on app/controllers/following_relations_controller.rb:
```diff
@@ -0,0 +1,15 @@
+  def create
+    user = User.find(params[:user_id])
+    current_user.follow(user)
+    redirect_to request.referer
```
明示的にリダイレクト先を指定せずに`request.referer`にしている理由があれば教えて欲しいです。

yocaji replied:
元のページに戻すには…という発想で入れたものでしたが、いただいたコメントを受けてもう一度調べて、この場面でば`user`で十分なことが分かりました🙏

maedana commented on app/views/users/show.html.erb:
```diff
@@ -1,5 +1,14 @@
+<div>
+  <%= link_to followings_user_path(@user) do %>
+    <span><%= t 'views.following.followings' %><%= @user.followings.count %></span>
+  <% end %>&nbsp;
```
可能な限り余白は`&nbsp;`など使わずcssで調整しましょうー

yocaji replied:
修正しました。

Commit:
レビューで指摘のあった点を修正した

- フォロー・アンフォロー時のリダイレクト先として users/:id を指定した
- フォロー関係モデルにバリデーションを追加した
- 既に存在しないフォロー関係をアンフォローしようとした時にエラーにならないようにした
- フォロー数・アンフォロー数表示のHTMLを修正した
- usersのルーティングに入っていた不要な指定を削除した
- フォロー・被フォローの列にnot null制約を追加した

maedana commented on app/controllers/following_relations_controller.rb:
```diff
@@ -4,12 +4,12 @@ class FollowingRelationsController < ApplicationController
     user = User.find(params[:user_id])
     current_user.follow(user)
-    redirect_to request.referer
+    redirect_to user
```
これでOKですが、より明示的にするなら`user_path(user)`となりますね。単に`user`を指定したらRails内部でよきに計らってくれるんですが、明示的に書くほうが後からコードを読む場合はリーダブルなのではないかなと思います。修正は任意で結構です。

maedana commented on db/migrate/20220408082907_create_following_relations.rb:
```diff
@@ -3,8 +3,8 @@
-      t.references :follower, foreign_key: { to_table: :users }
-      t.references :followed, foreign_key: { to_table: :users }
+      t.references :follower, foreign_key: { to_table: :users }, null: false
+      t.references :followed, foreign_key: { to_table: :users }, null: false
```
今回はこれでOKとしますが、もし本番環境等で既にこのmigrationが実行されている場合、元のmigrationファイルを直接直すのはNGです。`null: false`にするmigrationを別途作らないとだめですね。

Commit:
リダイレクト先の指定方法を修正した
